### PR TITLE
New CSP module and grep plugin

### PR DESCRIPTION
**Pull request by oxdef**
1. Re-factored CSP module: from functions to OOP (directive->policy->holder of policies) Now much easier to maintain the code and add new functionality 
2. Added retrieval of CSP policy from meta tag
3. Added new checks
4. Removed related to not standard  CSP realization. Now only modern CSPv2
5. Updated tests

### CODE REVIEW

**Comment by oxdef** [comment]
According to the log of CircleCI I don't understand how it is connected with this pull request :(

**Comment by andresriancho** [comment]
@oxdef shouldn't be your problem. I believe builds have been broken for some time now. CircleCI changes things, pip modules get outdated, etc.

File: w3af/core/controllers/csp/tests/test_csp_utils.py [hunk 1]
```diff
@@ -173,27 +109,17 @@ def test_provides_csp_features_no_case02(self):
...
-        #     img-src -> image-src
         header_value = "default-src ' '; img-src ' '"
-        hrds = {CSP_HEADER_W3C: header_value}.items()
-        csp_headers = Headers(hrds)
-        
-        http_response = HTTPResponse(200, '', csp_headers, self.url, self.url)
-                
-        self.assertFalse(provides_csp_features(http_response))
+        csp = CSPPolicy()
+        print csp.init_value(header_value), csp.directives
```
Comment (andresriancho) [inline w3af/core/controllers/csp/tests/test_csp_utils.py]: Remove print

File: w3af/core/controllers/csp/tests/test_csp_utils.py [hunk 2]
```diff
@@ -220,352 +142,119 @@ def test_provides_csp_features_yes_case02(self):
...
-        vulns = find_vulns(http_response)  
+        header_value = "default-src 'self';script-src 'self'; object-src 'self';" \
+                       "style-src 'self'; img-src 'self'; media-src 'self';" \
+                       "frame-src 'self'; font-src 'self'; sandbox;" \
+                       "form-action foo.com/myCtx/act; connect-src 'self';"\
+                       "plugin-types application/pdf;"
+        csp = CSPPolicy()
+        csp.init_value(header_value)
+        vulns = csp.find_vulns()
         self.assertEqual(len(vulns), 0)
```
Comment (andresriancho) [inline w3af/core/controllers/csp/tests/test_csp_utils.py]: I like these new tests, they are easier to read than the previous version.

File: w3af/plugins/grep/csp.py [hunk 3]
```diff
@@ -19,158 +19,163 @@
...
-                for csp_vuln in csp_vulns_list:        
-                    if csp_vuln.desc.strip().lower() == ref:
-                        if vuln_store_item.resp_id not in list_resp_id:
-                            list_resp_id.append(vuln_store_item.resp_id)
+        :return: A list of option objects for this plugin.
+        """
+        ol = OptionList()
 
-        return list_resp_id
+        d = 'Maximum number of sources in source list'
```
Comment (andresriancho) [inline w3af/plugins/grep/csp.py]: This description doesn't make much sense if you don't read the code.

 * Change it to something meaningful
 * Add a help

File: w3af/plugins/grep/csp.py [hunk 4]
```diff
@@ -19,158 +19,163 @@
...
+        o = opt_factory('source_limit', self._source_limit, d, 'integer')
+        ol.add(o)
 
+        d = 'Trusted hosts for finding untrusted ones in source list'
```
Comment (andresriancho) [inline w3af/plugins/grep/csp.py]: Same as above

File: w3af/plugins/grep/csp.py [hunk 5]
```diff
@@ -19,158 +19,163 @@
...
+
+    def get_long_desc(self):
         """
-        Implements method from base class.
+        :return: A DETAILED description of the plugin functions and features.
         """
-        return ['url','resp_id']
\ No newline at end of file
+        return """
+        This plugin identifies incorrect or too permissive Content Security Policy version 2
```
Comment (andresriancho) [inline w3af/plugins/grep/csp.py]: Is version 2 back compatible with CSP version 1? Maybe the description should say "too permissive Content Security Policy versions 1 and 2" ?

File: w3af/plugins/grep/csp.py [hunk 6]
```diff
@@ -19,158 +19,163 @@
...
+"""
+
+    vuln_without_tpl = """
+There were found URL without CSP. Please, review it:
+
+    %s
+"""
+
+    vuln_no_csp_tpl = """
+    The whole target has no CSP protection.
```
Comment (andresriancho) [inline w3af/plugins/grep/csp.py]: Unnecessary 4-spaces inside string?

**Comment by oxdef** [comment]
Andres, thanks for feedback! All has been fixed except:

> Is version 2 back compatible with CSP version 1? Maybe the description should say "too permissive Content Security Policy versions 1 and 2" ?

[It looks like](
https://www.w3.org/TR/CSP/#changes-from-level-1
) CSP2 is incompatible with Level 1. As I know [1.1](https://www.w3.org/TR/2014/WD-CSP11-20140211/) also contains directives which 2 doesn't contain.

Also added 2 more options for the plugin:

1. To report when policy doesn't contain report-uri
2. Manage if you want to report unsafe-eval

**Comment by oxdef** [comment]
Andres, do you have any other questions of feedback?

**Pull request closed without merging**